libbpf-tools/biostacks.bpf.c

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "biostacks.h"
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+#define MAX_ENTRIES 10240
+
+const volatile char targ_disk[DISK_NAME_LEN] = {};
+const volatile bool targ_ms = false;
+
+struct internal_rqinfo {
+	u64 start_ts;
+	struct rqinfo rqinfo;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct request *);
+	__type(value, struct internal_rqinfo);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} rqinfos SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct rqinfo);
+	__type(value, struct hist);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} hists SEC(".maps");
+
+static __always_inline bool disk_filtered(const char *disk)
+{
+	int i;
+
+	for (i = 0; targ_disk[i] != '\0' && i < DISK_NAME_LEN; i++) {
+		if (disk[i] != targ_disk[i])
+			return false;
+	}
+	return true;
+}
+
+static __always_inline
+int trace_start(struct internal_rqinfo *i_rqinfo, void *ctx,
+		struct request *rq, u32 pid, u64 ts)
+{
+	i_rqinfo->start_ts = ts;
+	i_rqinfo->rqinfo.pid = pid;
+	i_rqinfo->rqinfo.kern_stack_size =
+		bpf_get_stack(ctx, i_rqinfo->rqinfo.kern_stack,
+			sizeof(i_rqinfo->rqinfo.kern_stack), 0);
+	bpf_get_current_comm(&i_rqinfo->rqinfo.comm,
+			sizeof(&i_rqinfo->rqinfo.comm));
+	bpf_map_update_elem(&rqinfos, &rq, i_rqinfo, 0);
+	return 0;
+}
+
+SEC("fentry/blk_account_io_start")
+int BPF_PROG(fentry__blk_account_io_start, struct request *rq)
+{
+	u32 pid = bpf_get_current_pid_tgid();
+	struct internal_rqinfo i_rqinfo = {};
+	u64 ts = bpf_ktime_get_ns();
+
+	bpf_probe_read_kernel_str(&i_rqinfo.rqinfo.disk,
+				sizeof(i_rqinfo.rqinfo.disk),
+				rq->rq_disk->disk_name);
+	if (!disk_filtered(i_rqinfo.rqinfo.disk))
+		return 0;
+	return trace_start(&i_rqinfo, ctx, rq, pid, ts);
+}
+
+SEC("kprobe/blk_account_io_merge_bio")
+int BPF_KPROBE(kprobe__blk_account_io_merge_bio, struct request *rq)
+{
+	struct internal_rqinfo *i_rqinfop, i_rqinfo = {};
+	u32 pid = bpf_get_current_pid_tgid();
+	u64 ts = bpf_ktime_get_ns();
+
+	i_rqinfop = bpf_map_lookup_elem(&rqinfos, &rq);
+	if (!i_rqinfop) {
+		i_rqinfop = &i_rqinfo;
+		BPF_CORE_READ_STR_INTO(&i_rqinfop->rqinfo.disk, rq, rq_disk,
+				disk_name);
+		if (!disk_filtered(i_rqinfop->rqinfo.disk))
+			return 0;
+	}
+	return trace_start(i_rqinfop, ctx, rq, pid, ts);
+}
+
+SEC("fentry/blk_account_io_done")
+int BPF_PROG(fentry__blk_account_io_done, struct request *rq)
+{
+	u64 slot, ts = bpf_ktime_get_ns();
+	struct internal_rqinfo *i_rqinfop;
+	struct hist *histp, zero = {};
```

Comment:
this zero stuff is going to be zero-initialized at every call, it's a bit wasteful. Just declare it as static volatile variable once.

Reply:
Sure, thanks.